gossip/gossip.go

```diff
@@ -159,7 +159,20 @@ func New(rpcContext *rpc.Context, resolvers []resolver.Resolver) *Gossip {
 	}
 
 	// Add ourselves as a SystemConfig watcher.
-	g.is.registerCallback(KeySystemConfig, g.updateSystemConfig)
+	updateC, unregister := g.RegisterUpdateChannel(KeySystemConfig)
+	g.stopper.RunWorker(func() {
+		for {
+			select {
+			case n := <-updateC:
+				g.updateSystemConfig(n)
+				updateC <- n
```

Comment:
If we're going to require lockstep execution like this it would be simpler to use callbacks than channels (and reading and writing from the same channel is definitely not idiomatic). #2978 proposed channels with the expectation that we'd use channels in the most straightforward way without extra buffering; if that's not possible then I'd rather continue to use callbacks, but run them on a designated goroutine so they run in a more controlled order.

We may also be able to relax some of the constraints given in the PR message. For example, it's often better to allow a channel to block than to create an opportunity for unlimited buffering. Or we could figure out a way to apply the pattern from the previous gossip/channel change in which `make(chan struct{}, 1)` is used as a signal that something has changed but the current value is read separately (making it safe to discard messages on the channel).